radix-engine/src/engine/track.rs

```diff
@@ -444,6 +456,95 @@ impl<'s, S: ReadableSubstateStore> Track<'s, S> {
         }
     }
 
+    pub fn take_lock<A: Into<Address>>(&mut self, addr: A) -> Result<(), TrackError> {
+        let address = addr.into();
+        let maybe_value = self.up_substates.remove(&address.encode());
+        if let Some(value) = maybe_value {
+            self.borrowed_substates_2
+                .insert(address, RefCell::new(value));
+            return Ok(());
+        }
+
+        if self.borrowed_substates_2.contains_key(&address) {
+            return Err(TrackError::Reentrancy);
+        }
+
+        if let Some(substate) = self.substate_store.get_substate(&address.encode()) {
+            self.downed_substates.push(substate.phys_id);
+            let value = match address {
+                Address::Component(_) => {
+                    let component = scrypto_decode(&substate.value).unwrap();
+                    SubstateValue::Component(component)
+                }
+                Address::Resource(_) => {
+                    let resource_manager = scrypto_decode(&substate.value).unwrap();
+                    SubstateValue::Resource(resource_manager)
+                }
+                Address::Vault(..) => {
+                    let vault = scrypto_decode(&substate.value).unwrap();
+                    SubstateValue::Vault(vault)
+                }
+                _ => panic!("Attempting to borrow unsupported value"),
```

Comment:
Is the address anything that can be crafted by a user? How are we preventing this panic?